src/main/java/xyz/nucleoid/bedwars/game/active/BwBroadcast.java

```diff
@@ -23,76 +24,65 @@
     public void broadcastTrapSetOff(BwActive.TeamState team) {
         MutablePlayerSet players = team.players;
 
-        players.sendMessage(new LiteralText("A player set off your team trap!").formatted(Formatting.BOLD, Formatting.RED));
-        this.sendTitle(players, new LiteralText("Trap activated!").formatted(Formatting.RED), null);
+        players.sendMessage(new TranslatableText("message.bedwars.trap_set_off").formatted(Formatting.BOLD, Formatting.RED));
+        this.sendTitle(players, new TranslatableText("title.bedwars.trap_set_off").formatted(Formatting.RED), null);
         players.sendSound(SoundEvents.BLOCK_BELL_USE);
     }
 
-    public void broadcastTeamUpgrade(BwParticipant participant, Text message) {
-        ServerPlayerEntity player = participant.player();
-
-        Text broadcast;
-        if (player != null) {
-            broadcast = player.getDisplayName().shallowCopy().append(" ").append(message).formatted(Formatting.BOLD, Formatting.AQUA);
-        } else {
-            broadcast = new LiteralText("A player ").append(message).formatted(Formatting.BOLD, Formatting.AQUA);
-        }
-
-        this.game.playersFor(participant.team).sendMessage(broadcast);
+    public void broadcastToTeam(GameTeam team, MutableText upgradeText) {
+        this.game.playersFor(team).sendMessage(upgradeText);
     }
 
     public void broadcastGameOver(BwWinStateLogic.WinResult winResult) {
         GameTeam winningTeam = winResult.getTeam();
         if (winningTeam != null) {
             this.game.players().sendMessage(
-                    new LiteralText(winningTeam.getDisplay() + " team won the game!")
-                            .formatted(winningTeam.getFormatting(), Formatting.BOLD)
+                    new TranslatableText("message.bedwars.team_win", winningTeam.getDisplay()).formatted(winningTeam.getFormatting(), Formatting.BOLD)
             );
         } else {
-            this.game.players().sendMessage(new LiteralText("The game ended in a draw!").formatted(Formatting.BOLD));
+            this.game.players().sendMessage(new TranslatableText("message.bedwars.draw").formatted(Formatting.BOLD));
         }
     }
 
     public void broadcastDeath(ServerPlayerEntity player, ServerPlayerEntity killer, DamageSource source, boolean eliminated) {
         // TODO: we can do more specific messages in the future
-        MutableText announcement = player.getDisplayName().shallowCopy()
-                .append(new LiteralText(" was killed").formatted(Formatting.GRAY));
+        MutableText announcement;
 
-        if (killer != null) {
-            announcement = announcement.append(new LiteralText(" by ").formatted(Formatting.GRAY)).append(killer.getDisplayName());
+        if (killer == null) {
+            announcement = new TranslatableText("message.bedwars.player_death", player.getDisplayName().shallowCopy()).formatted(Formatting.GRAY);
+        }
+        else {
```

Comment:
This is completely pointless, and could lead to a NPE. You're checking if the killer is null and then getting the name of the killer in the else statement